ee/katc/table.go

```diff
@@ -0,0 +1,186 @@
+package katc
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"regexp"
+	"strings"
+
+	"github.com/osquery/osquery-go/plugin/table"
+)
+
+const sourceColumnName = "source"
+
+// katcTable is a Kolide ATC table. It queries the source and transforms the response data
+// per the configuration in its `cfg`.
+type katcTable struct {
+	cfg          katcTableConfig
+	columnLookup map[string]struct{}
+	slogger      *slog.Logger
+}
+
+// newKatcTable returns a new table with the given `cfg`, as well as the osquery columns for that table.
+func newKatcTable(tableName string, cfg katcTableConfig, slogger *slog.Logger) (*katcTable, []table.ColumnDefinition) {
+	columns := []table.ColumnDefinition{
+		{
+			Name: sourceColumnName,
+			Type: table.ColumnTypeText,
+		},
+	}
+	columnLookup := map[string]struct{}{
+		sourceColumnName: {},
+	}
+	for i := 0; i < len(cfg.Columns); i += 1 {
+		columns = append(columns, table.ColumnDefinition{
+			Name: cfg.Columns[i],
+			Type: table.ColumnTypeText,
+		})
+		columnLookup[cfg.Columns[i]] = struct{}{}
+	}
+
+	return &katcTable{
+		cfg:          cfg,
+		columnLookup: columnLookup,
+		slogger: slogger.With(
+			"table_name", tableName,
+			"table_type", cfg.SourceType,
+			"table_source", cfg.Source,
+		),
+	}, columns
+}
+
+// generate handles queries against a KATC table.
+func (k *katcTable) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	// Fetch data from our table source
+	dataRaw, err := k.cfg.SourceType.dataFunc(ctx, k.slogger, k.cfg.Source, k.cfg.Query, getSourceConstraint(queryContext))
+	if err != nil {
+		return nil, fmt.Errorf("fetching data: %w", err)
+	}
+
+	// Process data
+	transformedResults := make([]map[string]string, 0)
+	for _, s := range dataRaw {
+		for _, dataRawRow := range s.rows {
+			// Make sure source is included in row data
+			rowData := map[string]string{
+				sourceColumnName: s.path,
+			}
+
+			// Run any needed transformations on the row data
+			for _, step := range k.cfg.RowTransformSteps {
+				dataRawRow, err = step.transformFunc(ctx, k.slogger, dataRawRow)
+				if err != nil {
+					return nil, fmt.Errorf("running transform func %s: %w", step.name, err)
+				}
+			}
+
+			// After transformations have been applied, we can cast the data from []byte
+			// to string to return to osquery.
+			for key, val := range dataRawRow {
+				rowData[key] = string(val)
+			}
+			transformedResults = append(transformedResults, rowData)
+		}
+	}
+
+	// Now, filter data to ensure we only return columns in k.columnLookup
+	filteredResults := make([]map[string]string, 0)
+	for _, row := range transformedResults {
+		includeRow := true
+		filteredRow := make(map[string]string)
+		for column, data := range row {
+			if _, expectedColumn := k.columnLookup[column]; !expectedColumn {
+				// Silently discard the column+data
+				continue
+			}
+
+			filteredRow[column] = data
+
+			// No need to check the rest of the row
+			if !includeRow {
```

Comment:
how does `includeRow` become false?

Reply:
Oh, haha -- it's not actually needed anymore. This was leftover from when I was doing additional row filtering here. Updated to remove, thank you!